File: Data Structures & Algorithms/max-water-container/submission-3.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() - 1;

        int maxa = 0;
        while (l < r){
            int a = (r - l) * min(heights[r], heights[l]);
            maxa = max(maxa, a);
            if (heights[l] < heights[r]){
                l++;
            } else {
                r--;
            }

        }

        return maxa;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-3.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l = 0
        while l + 1 < len(height) and height[l + 1] >= height[l]:
            l += 1

        r = len(height) - 1
        while r - 1 >= 0 and height[r - 1] >= height[r]:
            r -= 1

        heighest = [0] * len(height)

        currMax = height[r]
        for i in range(r - 1, l, -1):
            currMax = max(currMax, height[i])
            heighest[i] = currMax

        print(heighest)
        accumulator = 0
        for i in range(l + 1, r):
            if height[i] >= height[l]:
                l = i
                continue
            accumulator += min(height[l], heighest[i]) - height[i]

        return accumulator